eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/PaymentOnion.scala

```diff
@@ -279,16 +253,16 @@ object PaymentOnion {
     val invoiceRoutingInfo = records.get[InvoiceRoutingInfo].map(_.extraHops)
   }
 
-  case class FinalTlvPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends FinalPayload with TlvFormat {
-    override val amount = records.get[AmountToForward].get.amount
-    override val expiry = records.get[OutgoingCltv].get.cltv
-    override val paymentSecret = records.get[PaymentData].get.secret
-    override val totalAmount = records.get[PaymentData].map(_.totalAmount match {
+  case class FinalPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends PerHopPayload with PaymentPacket with TrampolinePacket {
```

Comment:
I'm confused, how come all `FinalPayload` are `TrampolinePacket`?

Reply:
We use these traits (`PaymentPacket` and `TrampolinePacket`) as "this could be this packet type", not "it has to be this packet type". I agree that this is a bit confusing (it was already the case before but it was hidden by a more complex trait hierarchy) but it's quite handy! I'll try to think of an alternative to clean that up.

Reply:
This hierarchy is quite handy, because the way we use the `PacketType` is to restrict what kinds of payloads can be used in each context (trampoline or normal payment). I think it works well, but if you have a better alternative let me know!

Reply:
Maybe something like that?

```scala
  sealed trait FinalPayload extends PerHopPayload {
    val records: TlvStream[OnionPaymentPayloadTlv]
    val amount = records.get[AmountToForward].get.amount
    ...
    val paymentMetadata = records.get[PaymentMetadata].map(_.data)
  }

  case class PaymentFinalPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends FinalPayload with PaymentPacket

  case class TrampolineFinalPayload(records: TlvStream[OnionPaymentPayloadTlv]) extends FinalPayload with TrampolinePacket
```

Didn't dig much, and that would be for another PR anyway.

Reply:
Agreed, I'll dig into this more in a follow-up PR, along with the improvements in `Sphinx.scala`